libraries/lib-stretching-sequence/ClipSegment.cpp

```diff
@@ -42,10 +42,9 @@ sampleCount GetLastReadSample(
 sampleCount
 GetTotalNumSamplesToProduce(const ClipInterface& clip, double durationToDiscard)
 {
-   return sampleCount { (clip.GetVisibleSampleCount().as_double() -
-                         durationToDiscard * clip.GetRate()) *
-                           clip.GetStretchRatio() +
-                        .5 };
+   return sampleCount { clip.GetVisibleSampleCount().as_double() *
+                           clip.GetStretchRatio() -
+                        durationToDiscard * clip.GetRate() + .5 };
```

Comment:
I spent a surprising amount of time studying this small commit to convince myself it was correct.  Rereading the whole class led me to push one more commit for an possible undiscovered problem.  I think backwards scrubbing with stretch applied might be wrong.

Reply:
I found this small commit difficult to understand because `durationToDiscard` is multiplied by the clip's rate only in one place, but then also divided by stretch ratio in another place.

I think the members `mLastReadSample` and `mTotalNumSamplesToProduce` should be commented.

`mLastReadSample` counts raw samples but `mTotalNumSamplesToProduce` counts stretched.

Also, what should have made the bug "obvious" in earlier reviews -- is that the clip's rate should always be divided by the stretch ratio, never multiplied by it.

Reply:
I considered some time ago already providing a class only for the implementation of `TimeAndPitchSource` and let `ClipSegment` use it. It could have avoided this confusion. Now `mLastReadSample` is a member of the source, and represents a sample index before stretching. `mTotalNumSamplesToProduce` is a member of `ClipSegment` and represents a sample count after stretching.

Reply:
Now I did it. The commit "review comments: extract TimeAndPitchSource from ClipSegment ..." only just does that - no change in logic.